ldap/servers/slapd/back-ldbm/db-mdb/mdb_import_threads.c

```diff
@@ -77,121 +123,256 @@ dup_data(MDB_val *to, MDB_val *from)
     }
 }
 
-int
-dbmdb_import_workerq_init(ImportJob *job, ImportQueue_t *q, int slot_size, int max_slots)
+static inline int __attribute__((always_inline))
+cmp_data(MDB_val *d1, MDB_val *d2)
 {
-    q->job = job;
-    pthread_mutex_init(&q->mutex, NULL);
-    pthread_cond_init(&q->cv, NULL);
-    q->slot_size = slot_size;
-    q->max_slots = max_slots;
-    q->used_slots = 0;
-    q->slots = (WorkerQueueData_t*)slapi_ch_calloc(max_slots, slot_size);
-    return 0;
+    if (d1->mv_size != d2->mv_size) {
+        return d1->mv_size - d2->mv_size;
+    }
+    return memcmp(d1->mv_data, d2->mv_data, d1->mv_size);
 }
 
-void thread_abort(ImportWorkerInfo *info)
+/* Atomic compare and swap */
+static inline int __attribute__((always_inline))
+cas_pt(volatile void *addr, volatile void *old, volatile void *new)
+{
+    PR_ASSERT(sizeof (void*) == 8);
+    return __sync_bool_compare_and_swap_8((void**)addr, (int64_t)old, (int64_t)(new));
+}
+
+/* Tell whrether a job thread is asked to stop */
+static inline int __attribute__((always_inline))
+info_is_finished(ImportWorkerInfo *info)
+{
+    return (info->command == ABORT) ||
+           (info->command == STOP) ||
+           (info->state == ABORTED) ||
+           (info->state == FINISHED) ||
+           (info->job->flags & FLAG_ABORT);
+}
+
+void
+thread_abort(ImportWorkerInfo *info)
 {
     info->state = ABORTED;
 }
 
+void
+safe_cond_wait(pthread_cond_t *restrict cond, pthread_mutex_t *restrict mutex)
+{
+	struct timespec cvtimeout;
+    clock_gettime(CLOCK_REALTIME, &cvtimeout);
+    cvtimeout.tv_nsec += 100 * 1000 * 1000;
+    pthread_cond_timedwait(cond, mutex, &cvtimeout);
+}
 
-void safe_cond_wait(pthread_cond_t *restrict cond, pthread_mutex_t *restrict mutex)
+void
+wait_for_starting(ImportWorkerInfo *info)
 {
-        struct timespec cvtimeout;
-        clock_gettime(CLOCK_REALTIME, &cvtimeout);
-        cvtimeout.tv_nsec += 100 * 1000 * 1000;
-        pthread_cond_timedwait(cond, mutex, &cvtimeout);
+    PRIntervalTime sleeptime;
+    sleeptime = PR_MillisecondsToInterval(import_sleep_time);
+    /* pause until we're told to run */
+    while ((info->command == PAUSE) && !info_is_finished(info)) {
+        info->state = WAITING;
+        DS_Sleep(sleeptime);
+    }
+    info->state = RUNNING;
 }
 
+/***************************************************************************/
+/************************ ImportNto1Queue functions ************************/
+/***************************************************************************/
+
+/* generic code for a queue used by N provider threads to a single consumer */
+
+
+/* typical shouldwait callback for ImportNto1Queue */
 int
-dbmdb_import_workerq_push(ImportQueue_t *q, WorkerQueueData_t *data)
+generic_shouldwait(ImportNto1Queue_t *q)
 {
-    WorkerQueueData_t *slot =NULL;
+    ImportWorkerInfo *info = q->info;
+    return (q->nbitems < q->minitems && !info_is_finished(info));
```

Comment:
Setting the queue minimun to maxitem is a bad idea that mean that when pushing new data some the worker thread may have to wait for the writer thread empty the queue,
 I have done some tests with intermediary values but it did not improved the performance (in fact it slightly degrade it) 
 so IMHO we should better empty the queue and push the data towards the db as fast as we can ( i.e minitems == 1 )

Reply:
Thanks for the explanation. Do you mind to add the comment that minitems stick to 1 because tests shows it is better to write DB as soon as we have something to write.

Reply:
That is a good idea,  but I will do it in issue 5490 (because the queue code is refactored so I would rather limits the merge conflict. )
```
@@ -209,6 +201,7 @@ dbmdb_import_q_init(ImportNto1Queue_t *q, ImportWorkerInfo *info, int maxitems)
     pthread_cond_init(&q->cv, NULL);
     q->list = NULL;
     q->maxitems = maxitems;
+    /* Setting minitems to 1 because tests showed we get better performance with this value */
     q->minitems = 1;
```